File: benchmark_scripts/nbody-seq/jl/nbody.jl
using Random

DELTA_TIME=1.0
CONST_G=1.0

mutable struct Particle
    mass::Float64
    posx::Float64
    posy::Float64
    velx::Float64
    vely::Float64
    accx::Float64
    accy::Float64
end

function distance(p1::Particle, p2::Particle)::Float64
    return sqrt((p1.posx - p2.posx)^2 + (p1.posy - p2.posy)^2)
end

function update_position!(p::Particle)
    p.posx = p.posx + p.velx * DELTA_TIME;
    p.posy = p.posy + p.vely * DELTA_TIME; 
end

function update_velocity!(p::Particle)
    p.velx = p.velx + p.accx/p.mass * DELTA_TIME;
    p.vely = p.vely + p.accy/p.mass * DELTA_TIME; 
end

function update_acceleration!(p1::Particle, p2::Particle)
    dist = distance(p1, p2);
    f = CONST_G*(p1.mass * p2.mass)/(dist^2);

    p1.accx = p1.accx + f * (p2.posx - p1.posx) / dist 
    p1.accy = p1.accy + f * (p2.posy - p1.posy) / dist
end

function reset_force!(p::Particle)
    p.accx = 0.0
    p.accy = 0.0
end

function main()
    nb_particles = parse(Int64, ARGS[1])
    nb_timesteps = parse(Int64, ARGS[2])
    before = time()
    Random.seed!(0)
    particles = 1:nb_particles .|> (x -> Particle(rand(Float64, 1) * 100.0..., rand(Float64, 6)...))

    before_compute = time()
    for t in 1:nb_timesteps
        for i in 1:nb_particles
            for j in (i+1):nb_particles
                update_acceleration!(particles[i], particles[j])
                update_acceleration!(particles[j], particles[i])
            end
        end

        for i in 1:nb_particles
            update_velocity!(particles[i])
            update_position!(particles[i])
            reset_force!(particles[i])
        end

    end
    fin = time()
    duration_ete = fin - before
    duration_compute = fin - before_compute
    println("$(duration_ete), $(duration_compute), $(particles[1].posx)")
end

main()

File: benchmark_scripts/nbody-seq/cpp/nbody.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <chrono>
#include <iomanip>


#define DELTA_TIME 1.0
#define CONST_G 1.0

class Particle {
 public:
    double mass;
    double posx;
    double posy;
    double velx;
    double vely;
    double accx;
    double accy;

    double distance(Particle other);
    void update_position();
    void update_velocity();
    void update_acceleration(Particle other);
    void reset_force();

    Particle(double m, double x, double y, double vx, double vy, double ax, double ay) {
        mass = m;
        posx = x; posy = y;
        velx = vx; vely = vy;
        accx = ax; accy = ay;
    }
};

double Particle::distance(Particle other) {
    return sqrt(pow(posx - other.posx,2) + pow(posy - other.posy,2));
}

void Particle::update_position() {
    posx += velx * DELTA_TIME;
    posy += vely * DELTA_TIME;
}

void Particle::update_velocity() {
    velx += accx/mass * DELTA_TIME;
    vely += accy/mass * DELTA_TIME;
}

void Particle::update_acceleration(Particle other) {
    double dist = this->distance(other);
    double f = CONST_G*(mass * other.mass)/(pow(dist,2));

    accx += f * (other.posx - posx) / dist;
    accy += f * (other.posy - posy) / dist;
}

void Particle::reset_force() {
    accx = 0.0;
    accy = 0.0;
}

double _rand() {
    return (double) std::rand() / RAND_MAX;
}

Particle rand_particle() {
    return Particle(100.0 * _rand(), _rand(), _rand(), _rand(), _rand(), _rand(), _rand());
}


int main(int argc, char** argv) {
    // if (argc != 3) {
    //     fprintf(stderr, "USAGE: %s nb_particles nb_timesteps\n", argv[0]);
    //     return 1;
    // }
    int nb_particles = atoi(argv[1]);
    int nb_timesteps = atoi(argv[2]);

    auto begin = std::chrono::high_resolution_clock::now();

    std::srand(0);
    std::vector<Particle> particles = {};
    particles.reserve(nb_particles);
    // is there a more c++ way to do this ?
    for (int i = 0; i < nb_particles; i++) {
        particles.push_back(rand_particle());
    }

    auto begin_compute = std::chrono::high_resolution_clock::now();

    for (int ts = 1; ts <= nb_timesteps; ts++) {
        for (int i = 0; i < nb_particles; i++) {
            for (int j = i+1; j < nb_particles; j++) {
                particles[i].update_acceleration(particles[j]);
                particles[j].update_acceleration(particles[i]);
            }
        }
        for (int i = 0; i < nb_particles; i++) {
            particles[i].update_velocity();
            particles[i].update_position();
            particles[i].reset_force();
        }
    }
    auto end = std::chrono::high_resolution_clock::now();
    auto duration_compute = std::chrono::duration<float>(end - begin_compute);
    auto duration_ete = std::chrono::duration<float>(end - begin);
    std::cout << duration_ete.count() << "," << duration_compute.count() << "," << std::setprecision (16) << particles[0].posx << std::endl;
    return 0;
}
